src/main/cpp/VLFeat.cxx

```diff
@@ -239,33 +223,42 @@ JNIEXPORT jshortArray JNICALL Java_utils_external_VLFeat_getSIFTs (
     printf ("Error assigning memory for image buffer");
     exit(1);
   }
-  // calculate and get the denseSIFTdescriptors 
-  // NOTE! we need to clean up this array with free the passed container pointer. 
+  // calculate and get the denseSIFTdescriptors
+  // NOTE! we need to clean up this array with free the passed container pointer.
   DescSet* dSiftSet = getMultiScaleDSIFTs_f(&pim, pimData, step, bin, numScales, scaleStep);
   int numDesc = dSiftSet->numDesc;
   float* floatResult = dSiftSet->descriptors;
+  const VlDsiftKeypoint* keypoints = dSiftSet->keypoints;
   free (dSiftSet);
-  jshort* jshortResult = (jshort*) malloc(numDesc*dims*sizeof(jshort));
-
-  // transpose the descriptors. 
+  jint* jintResult = (jint*) malloc(numDesc*(dims+3)*sizeof(jint));
+  // transpose the descriptors.
   int binT = 8;
   int binX = 4;
   int binY = 4;
   float *tmpDescr = (float*) malloc(sizeof(float) * dims) ;
-  if (jshortResult != NULL) {  // loop throug and make unsigned dims and put in short to save memory
+  if (jintResult != NULL) {  // loop throug and make unsigned dims and put in int to save memory
     int currLoc = 0;
     for (int i=0; i<numDesc; i++) {
       vl_dsift_transpose_descriptor (tmpDescr, (floatResult + i*128), binT, binX, binY) ;
 
       for (int x=0; x<dims; x++) {
-        
+
         unsigned int v = (512 * tmpDescr[x]);
-        jshortResult[currLoc++] = ((v < 255) ? v : 255);
-        
+        jintResult[currLoc++] = ((v < 255) ? v : 255);
+
       } // end for x
+      if (memcmp((void*) (dSiftSet->keypoints + i), (void*) (&sentinel), sizeof(VlDsiftKeypoint))) {
+        jintResult[currLoc++] = (int) keypoints[i].x;
```

Comment:
I'm going to assume that you've verified by hand or by unit test that the results of these feature vectors are actually correct (that is, you've got the correct (x,y,scale) at the end of each sift descriptor.